### PR TITLE
Add `prefer-query-selector` rule

### LINKED ISSUE #171: Rule proposal: `prefer-query-selector`
Prefer `querySelector` over `getElementById`, `querySelectorAll` over `getElementsByClassName` and `getElementsByTagName`. They are not faster than `querySelector` and it's better to be consistent.

## Fail

```js
document.getElementById('foo');
```

## Pass

```js
document.querySelector('#foo');
```

### PR DESCRIPTION
**Pull request by janowsiany**
fixes #171

### CODE REVIEW

File: docs/rules/prefer-query-selector.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+```js
+document.getElementById('foo');
+```
+
+
+## Pass
+
+```js
+document.querySelector('#foo');
+```
```
Comment (sindresorhus) [inline docs/rules/prefer-query-selector.md]: Include some more examples in `Fail` and `Pass`.

File: index.js [hunk 2]
```diff
@@ -35,6 +35,7 @@ module.exports = {
 				'unicorn/import-index': 'error',
 				'unicorn/new-for-builtins': 'error',
 				'unicorn/regex-shorthand': 'error',
+				'unicorn/prefer-query-selector': 'error',
```
Comment (sindresorhus) [inline index.js]: Should be added to the end. You also need to add this to https://github.com/sindresorhus/eslint-plugin-unicorn#usage

File: rules/prefer-query-selector.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	'getElementById',
+	'querySelectorAll',
+	'getElementsByClassName',
+	'getElementsByTagName'
+];
+
+const VALID_QUOTES = /([',",`])/;
+const getRange = (prop, node) => [prop.start, node.arguments[0].end];
+const getReplacement = (ctx, identifierName, node) => {
+	const argAsTxt = ctx.getSourceCode().getText(node.arguments[0]);
```
Comment (sindresorhus) [inline rules/prefer-query-selector.js]: It would be better to operate on the AST nodes instead of the text. Then you don't have to deal with possible quotes, escapes, etc, it gets complicated quickly. Looks at how some of the other rules are implemented, for inspiration.

File: rules/prefer-query-selector.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+const forbiddenIdentifierNames = [
+	'getElementById',
+	'querySelectorAll',
+	'getElementsByClassName',
+	'getElementsByTagName'
+];
+
+const VALID_QUOTES = /([',",`])/;
+const getRange = (prop, node) => [prop.start, node.arguments[0].end];
+const getReplacement = (ctx, identifierName, node) => {
```
Comment (sindresorhus) [inline rules/prefer-query-selector.js]: `ctx` => `context`

Generally, use readable names instead of abbreviations.

File: test/prefer-query-selector.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	invalid: [
+		{
+			code: 'document.getElementById("foo");',
+			errors: [{message: 'Prefer `querySelector` over `getElementById`.'}],
+			output: 'document.querySelector("#foo");'
+		},
+		{
+			code: 'document.querySelectorAll("p");',
+			errors: [{message: 'Prefer `querySelector` over `querySelectorAll`.'}],
+			output: 'document.querySelector("p");'
```
Comment (sindresorhus) [inline test/prefer-query-selector.js]: This is incorrect. It should not prefer `querySelector` over `querySelectorAll`. Both are valid.

File: test/prefer-query-selector.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		},
+		{
+			code: 'document.getElementsByClassName("foo");',
+			errors: [{message: 'Prefer `querySelector` over `getElementsByClassName`.'}],
+			output: 'document.querySelector(".foo");'
+		},
+		{
+			code: 'document.getElementsByClassName("foo bar");',
+			errors: [{message: 'Prefer `querySelector` over `getElementsByClassName`.'}],
```
Comment (sindresorhus) [inline test/prefer-query-selector.js]: No need to repeat the error in every assertion. Just build them upfront, like done in https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/test/new-for-builtins.js

File: test/prefer-query-selector.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+			output: 'document.querySelector(".foo.bar");'
+		},
+		{
+			code: 'document.getElementsByTagName("div");',
+			errors: [{message: 'Prefer `querySelector` over `getElementsByTagName`.'}],
```
Comment (sindresorhus) [inline test/prefer-query-selector.js]: We should suggest `querySelectorAll` for `getElementsByTagName` and `getElementsByClassName`, since they can return multiple elements, and `querySelector` for `getElementById`.

File: test/prefer-query-selector.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		{
+			code: 'document.getElementById(\'foo\');',
+			errors: [{message: 'Prefer `querySelector` over `getElementById`.'}],
+			output: 'document.querySelector(\'#foo\');'
+		},
+		{
+			code: 'const qux = \'qux\'; document.getElementsByClassName(`foo ${qux}`);', // eslint-disable-line no-template-curly-in-string
+			errors: [{message: 'Prefer `querySelector` over `getElementsByClassName`.'}],
+			output: 'const qux = \'qux\'; document.querySelector(`.foo.${qux}`);' // eslint-disable-line no-template-curly-in-string
+		}
```
Comment (sindresorhus) [inline test/prefer-query-selector.js]: There needs to be some more tests:

- When a variable is used: `getElementsByClassName(foo)`.
- When a variable is partly used: `getElementsByClassName(foo + 'bar')`

**Comment by sindresorhus** [comment]
The rule needs to be added to the bottom of the list here: https://github.com/sindresorhus/eslint-plugin-unicorn#rules

**Comment by sindresorhus** [comment]
@futpib @MrHen Would you be able to help review? :)

File: test/prefer-query-selector.js (hunk 8 above)
Comment (futpib) [reply test/prefer-query-selector.js]: We should allow usage like `getElementById(id)` (when the argument is a variable), as to not to encourage writing `querySelector('#' + id)`.

In other words: 
`getElementById(id)`
is better then
`querySelector('#' + CSS.escape(id))`
is better then
`querySelector('#' + id)`

Same goes for `getElementsByClassName` and `getElementsByTagName`.

Comment (sindresorhus) [reply test/prefer-query-selector.js]: > We should allow usage like getElementById(id) (when the argument is not a string literal), as to not to encourage writing querySelector('#' + id).

I disagree. It's better to be consistent no matter what. It's an edge-case anyway, and `` querySelector(`#${id}`) `` is totally fine.

Why would you need to escape it with `querySelector` and not with `getElementById`?

Comment (futpib) [reply test/prefer-query-selector.js]: Why escape:
```js
const id = 'shoot foot';
const div = document.createElement('div');
div.id = id;
document.append(div);
document.querySelector(`#${CSS.escape(id)}`); // -> HTMLDivElement
document.querySelector(`#${id}`); // -> null
document.getElementById(id) // -> HTMLDivElement
```

Comment (sindresorhus) [reply test/prefer-query-selector.js]: I have done web dev for 10+ years and have yet to see an ID or class name with spaces. I honestly don't think we should care about such an obscure edge-case.

File: rules/prefer-query-selector.js (hunk 3 above)
Comment (janowsiany) [reply rules/prefer-query-selector.js]: Yes i know that operating on text is not the best. I wanted to do this by returning an array of fixes one for changing `MemberExpression` and the other for changing arguments `Literal`. However, although the documentation says that it is possible to return an array of fixes i could not make it work.

Comment (sindresorhus) [reply rules/prefer-query-selector.js]: Can you push those changes somewhere? Hard to tell what's wrong otherwise.

It should look like something like this:

```js
context.report({
    node: node,
    message: "Missing semicolon",
    fix: function(fixer) {
        return [
        	fixer.insertTextAfter(node, ";"),
        	fixer.insertTextAfter(node, ";")
        ]
    }
});
```

It should be an array of only `fixing` objects, which are returned by the `fixer` methods, like `fixer.insertTextAfter`.

File: rules/prefer-query-selector.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+};
+
+const create = context => {
+	return {
+		CallExpression(node) {
+			const {callee} = node;
+			const prop = callee.property;
+
+			const identifierName = prop.name;
+			const preferedSelector = forbiddenIdentifierNames[identifierName];
```
[the remark this answers is not in the document]
Comment (sindresorhus) [reply rules/prefer-query-selector.js]: A `Map` would be the correct data structure for this.

File: rules/prefer-query-selector.js (hunk 3 above)
Comment (janowsiany) [reply rules/prefer-query-selector.js]: Ok, @sindresorhus sorry if that takes too long but was investing this case with fixer returning an array. I believe it is because of our wrapper `eslint-ava-rule-tester` which has a dependency on old eslint package. I already made a PR and checked it locally. @jfmengels if you could review it here: https://github.com/jfmengels/eslint-ava-rule-tester/pull/11

**Comment by sindresorhus** [comment]
@janowsiany Ping :)

**Comment by janowsiany** [comment]
@sindresorhus sorry for such long delay, please revalidate

**Comment by sindresorhus** [comment]
This looks good to me now. Very nice work, @janowsiany 👍
